### ISSUE TITLE
jsonUpgraded flag implies upgradeAll flag

### ISSUE THREAD

**Issue by raffomania** [tjunnone/npm-check-updates#201]
When using the API, I have to enable the jsonUpgraded flag to get JSON as a response. When setting this flag to true, ncu also gives me updates that would usually be in the specified semver range, as if the upgradeAll flag was set to true.

**Comment by metaraine** [comment]
Thanks, the API should parallel the CLI.

**Comment by TheSavior** [comment]
It seems like even if specified via the CLI, the --jsonUpgraded flag will list all of the dependencies in the output JSON, not just the ones that are out of date. It doesn't appear that this is a api vs cli problem.

**Comment by raffomania** [comment]
Sure, this is less able bug than a feature request: I'd like the ability to get JSON output, but without automatically getting all updates as if I had enabled the `upgradeAll` flag.

**Comment by TheSavior** [comment]
@raffomania Is this behavior the inconsistency you are referring to? By specifying json, you get all packages not just the ones with new versions:

```
$ ncu

 eslint                  ^1.9.0  →   ^2.1.0 
 grunt-contrib-clean     ^0.6.0  →   ^1.0.0 
 grunt-eslint           ^17.2.0  →  ^18.0.0 
 grunt-karma             0.12.0  →   0.12.1 
 karma                   0.13.9  →  0.13.21 
 karma-browserify        ^4.4.0  →   ^5.0.1 
 karma-chrome-launcher    0.2.0  →    0.2.2 
 karma-mocha              0.2.0  →    0.2.2 
 nock                    ^3.1.0  →   ^7.2.1 

Run with -u to upgrade package.json

The following dependencies are satisfied by their declared version range, but the installed versions are behind. You can install the latest versions without modifying your package file by using npm update. If you want to update the dependencies in your package file anyway, use --upgradeAll.

 babel-preset-react    ^6.3.13  →   ^6.5.0 
 body-parser           ^1.14.1  →  ^1.15.0 
 grunt-express-server   ^0.5.1  →   ^0.5.2 
 jscs                   ^2.6.0  →  ^2.10.1
```

```
$ ncu --jsonUpgraded
{"app-module-path":"^1.0.5",
"babel-preset-react":"^6.5.0",
"body-parser":"^1.15.0",
"chai":"^3.5.0",
"eslint":"^2.1.0",
"eslint-plugin-react":"^3.16.1",
"express":"^4.13.4",
"grunt-contrib-clean":"^1.0.0",
"grunt-eslint":"^18.0.0",
"grunt-express-server":"^0.5.2",
"grunt-jscs":"^2.7.0",
"grunt-karma":"0.12.1",
"jscs":"^2.10.1",
"karma":"0.13.21",
"karma-browserify":"^5.0.1",
"karma-chrome-launcher":"0.2.2",
"karma-mocha":"0.2.2",
"load-grunt-tasks":"^3.4.0",
"nock":"^7.2.1",
"react":"^0.14.7",
"react-addons-test-utils":"^0.14.7",
"react-dom":"^0.14.7",
"request":"^2.69.0",
"serve-static":"^1.10.2",
"sinon":"^1.17.3"}
```

**Comment by carmelone** [comment]
How can I fix this?

Using this:

```
ncu.run({
    packageFile: 'package.json',
    silent: true,
    jsonUpgraded: true
}).then(function(upgraded) {
    console.log('dependencies to upgrade:', upgraded);
});
```

the result json contains one package which is ~0.1.1 in my node_modules, ~0.1.1 in my package.json and ~0.1.4 in the registry.

I think i should have 0 packages in my json results, right? Not this one, unless I set this option: `upgradeAll: true`, right?

**Comment by carmelone** [comment]
Any news? Can I help you fix this? Where do I begin to see in the code?

**Comment by raineorshine** [comment]
@carmelone Hi! Sorry for the delay, I was out of town the last two weeks.

Yes, the problem you describe is exactly what this issue is for. I have not had time to look at it myself unfortunately. It has something to do with the use of [jsonUpgraded](https://github.com/tjunnone/npm-check-updates/search?utf8=%E2%9C%93&q=jsonUpgraded&type=Code) and [upgradeAll](https://github.com/tjunnone/npm-check-updates/search?utf8=%E2%9C%93&q=upgradeAll&type=Code).

The first step towards fixing it would be to add a failing unit test. The `run` tests [are here](https://github.com/tjunnone/npm-check-updates/blob/186c13d6ada336f529cf75c8d0421d46c3321a18/test/test-ncu.js#L16-L46). It does not have full unit test coverage, hence the discrepancy with that and the CLI. This is not a difficult fix, it is just a little complicated since we need to think through the logical relationships of `upgradeAll` and `jsonUpgraded`. I'm not sure if need a separate flag, or if we need to modify the existing flag behavior.

If you are interesting in trying to implement a fix, I am happy to assist! Let me know if I can help explain anything else. The code base is not the worst and it's not the best.

**Comment by carmelone** [comment]
I don't know where to start.

**Comment by raineorshine** [comment]
@carmelone Maybe start by getting acquainted with the unit tests. Have you written unit tests before? Do the unit tests in this project make sense? [test-ncu](https://github.com/tjunnone/npm-check-updates/blob/186c13d6ada336f529cf75c8d0421d46c3321a18/test/test-ncu.js#L16-L46) is where the unit tests for the run method are currently contained. We'll need to add an `it` block that reproduces the offending behavior.

**Comment by raineorshine** [comment]
I've created a fix that maintains backwards-compatibility by only changing the behavior when `upgradeAll` is explicitly set to `false`. Try this and let me know if it works for you:

```
ncu.run({
    packageFile: 'package.json',
    silent: true,
    jsonUpgraded: true,
    upgradeAll: false
}).then(function(upgraded) {
    console.log('dependencies to upgrade:', upgraded);
});
```

**Comment by raineorshine** [comment]
@carmelone I have reopened the issue so you are welcome to post here. 

If I'm understanding you correctly from your last email (please correct me if I'm wrong):

Expected behavior: When using the Api, dependencies that have newer versions available for the package.json but are already installed in node_modules should be ignored by ncu, like the CLI. 

Actual behavior: 
When using the Api, dependencies that have newer versions available for the package.json but are already installed in node_modules *are not ignored by ncu*, but rather are suggested as updates. 

Is this correct?

**Comment by johnunclesam** [comment]
Yes, correct. Using `~` in version is ignored when using API.

**Comment by carmelone** [comment]
Yes @raineorshine. This is the big problem.

**Comment by raineorshine** [comment]
Okay, this helps a lot to understand the problem. I had misunderstood it before, thinking that it was just about the version numbers, but it is really about the presence of `node_modules` and looking at installed modules.

The explanation of the behavior is that when `packageFile` is specified, it just looks at the package.json in isolation, without looking for installed dependencies at all. It makes sense that it should have the same behavior as the CLI.

This is my top priority since it is important for several people, just know that this is an open source project that I do on the side, so I will have to tackle this as soon as I get some free time since I have other paid work commitments and I am in graduate school currently. I'm looking forward to getting this fixed, just want you to know I'm unable to provide the service of either a paid developer or a team of open source developers. Thanks for understanding!

**Comment by carmelone** [comment]
Any news?

**Comment by raineorshine** [comment]
I haven't had time yet this week, but I hope to do so very soon.

**Comment by raineorshine** [comment]
I added a `--packageFileDir` flag that when set will cause ncu to check for installed npm modules in the same directory as the specified `packageFile`. I added a test case so I know it works with the scenario I tested. (Confusingly, after I tested it it seemed that just setting `upgradeAll` to false gave the expected behavior, as provided in v.2.8.5. Maybe `packageFileDir` is redundant? I have not taken the time dig into the reasons for that. Regardless, the flag is in there so hopefully it does something!!!)

```
const upgraded = ncu.run({
    packageFile: 'package.json',
    packageFileDir: true,
    jsonUpgraded: true,
    upgradeAll: false
})
```

Published in v2.8.6. I (*really*) hope this works for you!

**Comment by carmelone** [comment]
Now, both with 2.8.6 and 2.8.7, I get `spawn npm ENOENT` error on Windows 10 x64.

😢

**Comment by raineorshine** [comment]
@carmelone That sounds like a different error. Would you creating a new issue for it? Is `npm` on your PATH?

**Comment by carmelone** [comment]
Yes it is. How to fix this? If I remove the line: `packageFileDir: true,` it works (but the bug is there also changing `upgradeAll` both true and false).

**Comment by raineorshine** [comment]
@carmelone Please report in separate issue as requested.

**Comment by carmelone** [comment]
https://github.com/tjunnone/npm-check-updates/issues/301

**Comment by carmelone** [comment]
Now `spawn error` on Windows is gone, but still I get updates to do also if I have already updates done in my node_modules.

Again, my need are:

If in my package.json I have "my-package": "~1.0.0" and `ncu.run` detects new version, let's say: "1.0.2" and with `npm update` I got it instaleld, the next `ncu.run` should not warning me about "my-package": "~1.0.0" to update because in my node_modules/my-package folder I already have 1.0.2 version.

How to accomplish this?

:(

**Comment by raineorshine** [comment]
Fixed in branch `v3` and will be released soon. See #471 for roadmap.
